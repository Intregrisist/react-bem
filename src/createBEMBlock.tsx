import * as PT from 'prop-types'
import * as React from 'react'

import { getDisplayName, isFunction } from './helpers'
import { BEMBlockClass, BEMBlockProps } from './types'
import resolveBEMNode, { ReactRenderResult } from './resolveBEMNode'

/**
 * Wraps a class with BEM block functionality, providing the BEM block name
 * via context and converting block and modifiers attributes into className
 * attributes.
 * @param ComponentClass The class to wrap with BEM block functionality.
 */
export default function createBEMBlock(
	ComponentClass: BEMBlockClass | React.ComponentClass,
) {

	return class Wrapped<P = {}, S = {}>
	extends (ComponentClass as BEMBlockClass)<P & BEMBlockProps, S> {

		static displayName = `BEMBlock(${getDisplayName(ComponentClass)})`

		static childContextTypes = {
			...ComponentClass.childContextTypes,
			block: PT.string.isRequired,
		}

		getChildContext() {
			return {
				...(isFunction(super.getChildContext) && (
					super.getChildContext.call(this))
				),
				block: this.props.block,
			}
		}

		render() {
			const {
				block,
				modifiers,
			} = this.props
			const rendered = super.render.call(this)
			return rendered && resolveBEMNode(rendered, {
				block: block as string,
				modifiers,
			}) as ReactRenderResult
		}
	}
}
