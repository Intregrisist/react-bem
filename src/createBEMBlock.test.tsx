import { render } from 'enzyme'
import * as PT from 'prop-types'
import * as React from 'react'

import createBEMBlock from './createBEMBlock'

describe('createBEMBlock', () => {

	it('extends input class with a displayName of "BEMBlock(Foo)"', () => {
		class Foo extends React.Component {
			render() {
				return null
			}
		}
		const FooBlock = createBEMBlock(Foo)
		expect(FooBlock.displayName).toEqual('BEMBlock(Foo)')
	})

	it('extends input propTypes with a block validator', () => {
		class Foo extends React.Component {
			static propTypes = {
				bar: PT.string,
			}
			render() {
				return null
			}
		}
		const FooBlock = createBEMBlock(Foo)
		expect(FooBlock.propTypes).toMatchSnapshot()
	})

	it('provides BEM block name via context', () => {
		class Foo extends React.Component {
			render() {
				return (
					<div>
						{this.props.children}
					</div>
				)
			}
		}
		class FooItem extends React.Component {
			render() {
				expect(this.props).toMatchSnapshot()
				return null
			}
		}
		const FooBlock = createBEMBlock(Foo)
		render(
			<FooBlock block="foo">
				<FooItem />
			</FooBlock>,
		)
	})

	it('merges modifiers prop with root node\'s modifiers', () => {
		class Foo extends React.Component {
			render() {
				return (
					<div modifiers="mod1" />
				)
			}
		}
		const FooBlock = createBEMBlock(Foo)
		expect(render(
			<FooBlock block="foo" modifiers="mod2" />,
		)).toMatchSnapshot()
	})
})
