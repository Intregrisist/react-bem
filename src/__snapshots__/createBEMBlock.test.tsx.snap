// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBEMBlock extends getChildContext() with the block value from props 1`] = `
Object {
  "bar": "baz",
  "block": "context-block",
}
`;

exports[`createBEMBlock merges modifiers prop with root node's modifiers 1`] = `
<div
  class="foo foo--mod1 foo--mod2"
/>
`;

exports[`createBEMBlock provides BEM block name via context 1`] = `
Object {
  "block": "context-block",
}
`;
